### PR TITLE
ENH: Zivot-Andrews unit-root test (issue 4746)

### PR DESCRIPTION
**Pull request by JimVaranelli**
Initial commit for Zivot-Andrews structural-break unit-root test (issue 4746).

### CODE REVIEW

**Comment by JimVaranelli** [comment]
Even after ensuring my fork was synced I still get this merge error. Very annoying.

**Comment by josef-pkt** [comment]
my guess is that you didn't branch off current master. (github doesn't show the parent of a branch or commit anymore.)
Maybe you started with an outdated local master checkout, or/and didn't `fetch` from github in some time.

BTW it is better to also add the topic to the branch name instead of just the issue number. That makes it easier to keep track of branches locally.
e.g. my local branch of the White PR us `white_spec_5452_rebased`

**Comment by josef-pkt** [comment]
Checking the history of this branch locally:
The branch of this PR is based on master from May 29, 2018, the same as the initial White spec PR

**Comment by JimVaranelli** [comment]
OK I'll use the more informative branch name. I forced the most recent checkout of the two conflict files and merged in my changes and pushed the commit but I am still seeing the merge conflict. And the 'Resolve Conflicts' button just spins and never shows the conflict.

**Comment by josef-pkt** [comment]
To resolve the merge conflicts and update to current master we use rebase. We avoid merging master into a feature branch because it makes the commit history messy and more difficult to work with.

`>git rebase remotes/origin/master`

where `origin` is my shortcut to statsmodels/statsmodels official repo

**Comment by josef-pkt** [comment]
Also we are not using `@` for `dot` yet.  (It will be available after dropping 2.7 and 3.4)

**Comment by JimVaranelli** [comment]
OK I will change the @ back to dot.

I rebased but it is still showing the conflict here???????

**Comment by JimVaranelli** [comment]
It looks like the merge conflict is finally fixed with the @/dot change. The issue most likely traces back to cloning the incorrect version of master (5/29/18 instead of 1/9/19). What is the correct command to use when cloning a new branch to ensure the most recent commits are included?

**Comment by JimVaranelli** [comment]
Finally got a clean build. The appveyor failure appears to be unrelated to this PR. Let me know if there is anything further I should do.

File: statsmodels/tsa/stattools.py [hunk 1]
```diff
@@ -1620,3 +1620,279 @@ def _sigma_est_kpss(resids, nobs, lags):
         resids_prod = np.dot(resids[i:], resids[:nobs - i])
         s_hat += 2 * resids_prod * (1. - (i / (lags + 1.)))
     return s_hat / nobs
+
+
+class ur_za(object):
```
Comment (jbrockmendel) [inline statsmodels/tsa/stattools.py]: Any reason not to use standard CamelCase naming?

File: statsmodels/tsa/stattools.py [hunk 2]
```diff
@@ -1620,3 +1620,279 @@ def _sigma_est_kpss(resids, nobs, lags):
...
+              (42.500, -4.11946), (45.000, -4.08064), (47.500, -4.04286),
+              (50.000, -4.00489), (52.500, -3.96837), (55.000, -3.93200),
+              (57.500, -3.89496), (60.000, -3.85577), (65.000, -3.77795),
+              (70.000, -3.69794), (75.000, -3.61852), (80.000, -3.52485),
+              (85.000, -3.41665), (90.000, -3.28527), (92.500, -3.19724),
+              (95.000, -3.08769), (96.000, -3.03088), (97.000, -2.96091),
+              (98.000, -2.85581), (99.000, -2.71015), (99.900, -2.28767), )
+        self.__za_critical_values['ct'] = np.asarray(self.__ct)
+
+    def __za_crit(self, stat, model='c'):
```
Comment (jbrockmendel) [inline statsmodels/tsa/stattools.py]: Double underscores aren’t used often.  Is there a special reason to do so here?

File: statsmodels/tsa/stattools.py [hunk 3]
```diff
@@ -1620,3 +1620,279 @@ def _sigma_est_kpss(resids, nobs, lags):
...
+        pval = crit[0]
+        cvdict = crit[1]
+        return zastat, pval, cvdict, baselags, bpidx
+
+    def __call__(self, x, trim=0.15, maxlag=None, regression='c',
+                 autolag='AIC'):
+        return self.run(x, trim=trim, maxlag=maxlag, regression=regression,
+                        autolag=autolag)
+
+za = ur_za()
```
Comment (jbrockmendel) [inline statsmodels/tsa/stattools.py]: Please give this a longer, more informative name.

If the class is t necessary, it’s fine to just implement this as a function.

File: statsmodels/tsa/stattools.py (hunk 1 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: To be perfectly honest I don't recall why I did it this way other than the fact that the the actual call is all lower (i.e. 'za'). More than happy to change it to CC (i.e. class ZivotAndrews).

File: statsmodels/tsa/stattools.py (hunk 2 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: Pretty sure I added the double underscore in case there was ever a desire to break this out into a general class with potential subclass inheritance. Although certainly the class name doesn't need the double underscore here. I'll change the doubles to single.

File: statsmodels/tsa/stattools.py (hunk 3 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: See first comment. I can call the class ZivotAndrewsUnitRoot and the actual call ZivotAndrews? I like the idea of keeping as a class due to the two ZA-specific helpers but can certainly break out if that is the preference. Let me know and I will update the PR as requested.

File: statsmodels/tsa/stattools.py (hunk 1 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: Changed the class name to ZivotAndrewsUnitRoot.

File: statsmodels/tsa/stattools.py (hunk 2 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: Replaced the double underscores ('__') with single underscores ('_').

File: statsmodels/tsa/stattools.py (hunk 3 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: Changed function call name from za() to zivot_andrews (following leinson_durbin naming convention).

**Comment by JimVaranelli** [comment]
Fixed the Codacy white space issue. Should have a clean merge now.

**Comment by josef-pkt** [comment]
this might still need a clean rebase on current master (with possibly some squashing)
I didn't check locally how the commits look like.

File: statsmodels/tsa/stattools.py [hunk 4]
```diff
@@ -1735,6 +1735,282 @@ def _sigma_est_kpss(resids, nobs, lags):
...
+             (50.000, -3.25316), (52.500, -3.21244), (55.000, -3.17124),
+             (57.500, -3.13211), (60.000, -3.09204), (65.000, -3.01135),
+             (70.000, -2.92897), (75.000, -2.83614), (80.000, -2.73893),
+             (85.000, -2.62840), (90.000, -2.49611), (92.500, -2.41337),
+             (95.000, -2.30820), (96.000, -2.25797), (97.000, -2.19648),
+             (98.000, -2.11320), (99.000, -1.99138), (99.900, -1.67466), )
+        self._za_critical_values['t'] = np.asarray(self._t)
+        # constant + trend model
+        self._ct = ((0.001, -38.17800), (0.100, -6.43107), (0.200, -6.07279),
+              (0.300, -5.95496), (0.400, -5.86254), (0.500, -5.77081),
```
Comment (jbrockmendel) [inline statsmodels/tsa/stattools.py]: check indentation.  if you're not familiar with the convention, you can run `flake8 --isolated statsmodels/tsa/stattools.py` and ignore all the alerts for the rest of the module.

Comment (jbrockmendel) [reply statsmodels/tsa/stattools.py]: are these actually needed as attributes of `self`?

If these attributes are identical across all instances and never edited, no need to bother with `__init__`, just put them as class attributes

File: statsmodels/tsa/stattools.py [hunk 5]
```diff
@@ -1735,6 +1735,282 @@ def _sigma_est_kpss(resids, nobs, lags):
...
+        """
+        table = self._za_critical_values[model]
+        y = table[:, 0]
+        x = table[:, 1]
+        # ZA cv table contains quantiles multiplied by 100
+        pvalue = np.interp(stat, x, y) / 100.0
+        cv = [1.0, 5.0, 10.0]
+        crit_value = np.interp(cv, y, x)
+        cvdict = {"1%" : crit_value[0], "5%" : crit_value[1],
+                  "10%" : crit_value[2]}
```
Comment (jbrockmendel) [inline statsmodels/tsa/stattools.py]: no space before colon

are there other names besides x, y that might be more informative?  one-letter names make it tough to search a file for all-places-this-is-used

File: statsmodels/tsa/stattools.py [hunk 6]
```diff
@@ -1735,6 +1735,282 @@ def _sigma_est_kpss(resids, nobs, lags):
...
+                exog[:(cutoff-1), 2] = 0
+                exog[(cutoff-1):, 2] = t_const[0:(nobs - bp + 1)]
+            # check exog rank on first iteration
+            if bp == start_period + 1:
+                o = OLS(dy[baselags:], exog, hasconst=1).fit()
+                if o.df_model < exog.shape[1] - 1:
+                    raise ValueError(
+                        'ZA: auxiliary exog matrix is not full rank.\n \
+                        cols (exc intercept) = {}  rank = {}'.format(
+                            exog.shape[1] - 1, o.df_model))
```
Comment (jbrockmendel) [inline statsmodels/tsa/stattools.py]: using modern string formatting here is nice, can you do the same above?

Also can you avoid the backslash and instead put single-quotes on both lines

File: statsmodels/tsa/stattools.py [hunk 7]
```diff
@@ -1735,6 +1735,282 @@ def _sigma_est_kpss(resids, nobs, lags):
...
+        """
+        if regression not in ['c', 't', 'ct']:
+            raise ValueError(
+                'ZA: regression option \'%s\' not understood' % regression)
+        if not isinstance(trim, float) or trim < 0 or trim > (1. / 3.):
+            raise ValueError(
+                'ZA: trim value must be a float in range [0, 0.333]')
+        x = np.asarray(x)
+        if x.ndim > 2 or (x.ndim == 2 and x.shape[1] != 1):
+            raise ValueError(
```
Comment (jbrockmendel) [inline statsmodels/tsa/stattools.py]: quick tests for the failure modes?

File: statsmodels/tsa/stattools.py [hunk 8]
```diff
@@ -1735,6 +1735,282 @@ def _sigma_est_kpss(resids, nobs, lags):
...
+                exog[cutoff:, 1] = c_const
+                exog[:, 2] = t_const[(baselags + 2):(nobs + 1)]
+                if regression == 'ct':
+                    exog[:cutoff, 3] = 0
+                    exog[cutoff:, 3] = t_const[1:(nobs - bp + 1)]
+            else:
+                exog[:, 1] = t_const[(baselags + 2):(nobs + 1)]
+                exog[:(cutoff-1), 2] = 0
+                exog[(cutoff-1):, 2] = t_const[0:(nobs - bp + 1)]
+            # check exog rank on first iteration
```
Comment (jbrockmendel) [inline statsmodels/tsa/stattools.py]: this method is doing a lot.  are there any reasonable sub-scopes it could be split into?

File: statsmodels/tsa/tests/test_stattools.py [hunk 9]
```diff
@@ -879,13 +879,55 @@ def test_innovations_algo_filter_kalman_filter(reset_randomstate):
     res = mod.filter(np.r_[ar_params, ma_params, sigma2])
 
     # Test that the two approaches are identical
+    assert_allclose(u, res.forecasts_error[0])
+    assert_allclose(theta[1:, 0], res.filter_results.kalman_gain[0, 0, :-1])
+    assert_allclose(llf_obs, res.llf_obs)
```
Comment (jbrockmendel) [inline statsmodels/tsa/tests/test_stattools.py]: im confused as to where these lines came from.  are these related to zivot_andrews?

File: statsmodels/tsa/tests/test_stattools.py [hunk 10]
```diff
@@ -879,13 +879,55 @@ def test_innovations_algo_filter_kalman_filter(reset_randomstate):
...
     assert_allclose(theta[1:, 0], res.filter_results.kalman_gain[0, 0, :-1],
                     atol=atol)
     assert_allclose(llf_obs, res.llf_obs, atol=atol)
 
 
+def test_zivot_andrews():
+    cur_dir = os.path.abspath(os.path.dirname(__file__))
+    resdir = os.path.join(cur_dir, "results")
+    zafiles = ['rgnp.csv', 'gnpdef.csv', 'stkprc.csv', 'rgnpq.csv', 'rand10000.csv']
+    for file in zafiles:
```
Comment (jbrockmendel) [inline statsmodels/tsa/tests/test_stattools.py]: file --> filename

File: statsmodels/tsa/tests/test_stattools.py [hunk 11]
```diff
@@ -879,13 +879,55 @@ def test_innovations_algo_filter_kalman_filter(reset_randomstate):
...
+            assert_almost_equal(res[0], -3.02761, decimal=3)
+            assert_almost_equal(res[1], 0.63993, decimal=3)
+            assert_equal(res[3], 12)
+            assert_equal(res[4], 102)
+        else:
+            res = zivot_andrews(mdl, regression='c', autolag='t-stat')
+            assert_almost_equal(res[0], -3.48223, decimal=3)
+            assert_almost_equal(res[1], 0.69111, decimal=3)
+            assert_equal(res[3], 25)
+            assert_equal(res[4], 7071)
```
Comment (jbrockmendel) [inline statsmodels/tsa/tests/test_stattools.py]: these should probably be 5 separate tests in a test class

for scalars `assert_equal(res[3], 25)` can just be `assert res[3] == 25`

File: statsmodels/tsa/tests/test_stattools.py [hunk 12]
```diff
@@ -879,13 +879,55 @@ def test_innovations_algo_filter_kalman_filter(reset_randomstate):
...
+        elif file == 'rgnpq.csv':
+            res = zivot_andrews(mdl, maxlag=12, regression='t', autolag='t-stat')
+            assert_almost_equal(res[0], -3.02761, decimal=3)
+            assert_almost_equal(res[1], 0.63993, decimal=3)
+            assert_equal(res[3], 12)
+            assert_equal(res[4], 102)
+        else:
+            res = zivot_andrews(mdl, regression='c', autolag='t-stat')
+            assert_almost_equal(res[0], -3.48223, decimal=3)
+            assert_almost_equal(res[1], 0.69111, decimal=3)
```
Comment (jbrockmendel) [inline statsmodels/tsa/tests/test_stattools.py]: im pretty sure assert_allclose is preferred to assert_almost_equal

File: statsmodels/tsa/tests/test_stattools.py (hunk 9 above)
Comment (JimVaranelli) [reply statsmodels/tsa/tests/test_stattools.py]: Hmmmm. This one looks related to the mangled merge. Must have missed these three lines during the rebase. Will fix this first.

Comment (JimVaranelli) [reply statsmodels/tsa/tests/test_stattools.py]: Fixed this issue along with the simple 'filename' change below. Will wait to make sure there are no further merge problems before starting on the rest of these.

File: statsmodels/tsa/tests/test_stattools.py (hunk 10 above)
Comment (JimVaranelli) [reply statsmodels/tsa/tests/test_stattools.py]: Done.

**Comment by JimVaranelli** [comment]
Looks like the merge should be OK now. Not sure why the Mac 2.7 pytest failed as the log was empty. I'll crank the rest of these out today.

File: statsmodels/tsa/stattools.py (hunk 4 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: 1) I fixed the indentation although there seems to be some ambiguity between the different style parsers. The form I used (which seems to be the most preferred) is:
list = (
  (x1,y1), ...
  (x2,y2), ...
)

2) I originally used __init__ for the CV list specifically in case there is ever a desire/need to split out the routine as a base class then inherit for different distributions. The question of class attribute vs object attribute is not too important here since the use model (in a single-threaded context) is to instantiate once then re-use the object for multiple runs as opposed to instantiating a new object for each run. I left it as is for now but can certainly update if there is a strong preference.

File: statsmodels/tsa/stattools.py (hunk 5 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: Fixed the colon spacing and renamed to (pcnts, stats) rather than (x, y).

File: statsmodels/tsa/stattools.py (hunk 6 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: Updated the other string formatting commands and removed the '\' as above.

File: statsmodels/tsa/stattools.py (hunk 7 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: Added failure mode tests to test_stattools.py.

File: statsmodels/tsa/stattools.py (hunk 8 above)
Comment (JimVaranelli) [reply statsmodels/tsa/stattools.py]: Split off the endog/exog creation/updating into two separate helper functions.

File: statsmodels/tsa/tests/test_stattools.py (hunk 11 above)
Comment (JimVaranelli) [reply statsmodels/tsa/tests/test_stattools.py]: Changed to a test class and switched to assert_allclose (see below).

File: statsmodels/tsa/tests/test_stattools.py (hunk 12 above)
Comment (JimVaranelli) [reply statsmodels/tsa/tests/test_stattools.py]: Now using assert_allclose.

**Comment by JimVaranelli** [comment]
Fixed the Codacy whitespace issue. Very surprised the IDE didn't catch this one.
